File: erlang/sieve/src/sieve.erl
-module(sieve).

-export([primes/1]).

%% @doc Compute the list of primes from 2 up to a given number.
-spec primes(Limit :: pos_integer()) -> [Primes :: pos_integer()].
primes(Limit) when Limit < 2 ->
    [];
primes(Limit) ->
    primes(2, round(math:sqrt(Limit)), [], lists:seq(3, Limit, 2)).

-spec primes(Prime, Limit, Primes, Candidates) -> Primes when
    Prime :: pos_integer(),
    Limit :: pos_integer(),
    Primes :: [pos_integer()],
    Candidates :: [pos_integer()].
primes(Prime, Limit, Primes, Candidates) when Prime > Limit ->
    lists:reverse([Prime | Primes]) ++ Candidates;
primes(Prime, Limit, Primes, Candidates) ->
    [NextPrime | NextCandidates] = [N || N <- Candidates, N rem Prime =/= 0],
    primes(NextPrime, Limit, [Prime | Primes], NextCandidates).

File: go/sieve/sieve.go
package sieve

import (
	"math"
	"sort"
)

func Sieve(limit int) []int {
	if limit <= 1 {
		return nil
	}

	sieve := make(map[int]struct{}, limit-1)
	for i := 2; i <= limit; i++ {
		sieve[i] = struct{}{}
	}

	sqrtLimit := int(math.Ceil(math.Sqrt(float64(limit))))
	for i := 2; i <= sqrtLimit; i++ {
		if _, present := sieve[i]; present {
			for j := i * i; j <= limit; j += i {
				delete(sieve, j)
			}
		}
	}

	keys := make([]int, 0, len(sieve))
	for k := range sieve {
		keys = append(keys, k)
	}

	sort.SliceStable(keys, func(i, j int) bool {
		return keys[i] < keys[j]
	})

	return keys
}
